sklearn/preprocessing/_discretization.py

```diff
@@ -203,10 +211,11 @@ def fit(self, X, y=None):
         if 'onehot' in self.encode:
             self._encoder = OneHotEncoder(
                 categories=[np.arange(i) for i in self.n_bins_],
-                sparse=self.encode == 'onehot')
+                sparse=self.encode == 'onehot',
+                dtype=output_dtype)
             # Fit the OneHotEncoder with toy datasets
             # so that it's ready for use after the KBinsDiscretizer is fitted
-            self._encoder.fit(np.zeros((1, len(self.n_bins_)), dtype=int))
+            self._encoder.fit(np.zeros((1, len(self.n_bins_))))
```

Comment:
I think we still may want this, not sure.

Reply:
It is the dtype of the toy dataset used to fit  the encoder. It only influences the dtype of the private attributes of OneHotEncoder when we fit it (if I am right). When we apply `self._encoder.transform(Xt)` (ie. `OneHotEncoder.transform(Xt)`), the output is cast with the dtype parameter of the encoder. So `dtype=int` seems useless here. Plus, the tests are still green. Do you want it back ?

Reply:
It might be best to pass `dtype=X.dtype` for consistency